lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -744,12 +745,116 @@ swift::castValueToABICompatibleType(SILBuilder *builder, SILPassManager *pm,
               false};
     }
   }
+  NominalTypeDecl *srcNominal = srcTy.getNominalOrBoundGenericNominal();
+  NominalTypeDecl *destNominal = destTy.getNominalOrBoundGenericNominal();
+  if (srcNominal && srcNominal == destNominal &&
+      !layoutIsTypeDependent(srcNominal) &&
+      srcTy.isObject() && destTy.isObject()) {
+    return {builder->createUncheckedForwardingCast(loc, value, destTy), false};
+  }
 
   llvm::errs() << "Source type: " << srcTy << "\n";
   llvm::errs() << "Destination type: " << destTy << "\n";
   llvm_unreachable("Unknown combination of types for casting");
 }
 
+namespace {
+  class TypeDependentVisitor : public CanTypeVisitor<TypeDependentVisitor, bool> {
```

Comment:
Can you put this in a shared utility instead of copy and pasting it from IRGen?

Reply:
This utility is for a different purpose and also behaves substantially different. It's not a copy-paste from IRGen.
What could be done, though is the extract the visit logic for stored properties and enum cases into a base class visitor, like `ValuePropertyTypeVisitor` from which both utilities can derive.
But I'd like to do that in a follow-up PR to minimize the risk for cherry-picking this PR to 6.0.